OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs

```diff
@@ -54,51 +51,57 @@ void IRulesetLoaded<WeaponInfo>.RulesetLoaded(Ruleset rules, WeaponInfo info)
 				VictimScanRadius = Util.MinimumRequiredVictimScanRadius(rules);
 		}
 
+		readonly string[] targetTypeAir = new string[] { "Air" };
+
 		public ImpactType GetImpactType(World world, CPos cell, WPos pos, Actor firedBy)
 		{
-			// Missiles need a margin because they sometimes explode a little above ground
-			// due to their explosion check triggering slightly too early (because of CloseEnough).
-			// TODO: Base ImpactType on target altitude instead of explosion altitude.
-			var airMargin = new WDist(128);
-
 			// Matching target actor
-			if (ValidImpactTypes.HasFlag(ImpactType.TargetHit) && GetDirectHit(world, cell, pos, firedBy, true))
-				return ImpactType.TargetHit;
+			if (VictimScanRadius > WDist.Zero)
+			{
+				var targetType = GetDirectHitTargetType(world, cell, pos, firedBy, true);
+				if (targetType == ImpactTargetType.ValidActor)
+					return ImpactType.TargetHit;
+				else if (targetType == ImpactTargetType.InvalidActor)
```

Comment:
Style nit: you can omit the `else` here (use `if` by itself) because the first branch returns.